rust/data_structures/bloom_filter.rs

```diff
@@ -0,0 +1,86 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+
+#[derive(Debug)]
+struct BloomFilter {
+    bits: Vec<bool>,
+    hash_functions: usize,
+}
+
+
+impl BloomFilter {
+
+    /// Creates a new BloomFilter which can contain up to size items
+    /// with a false positive rate less than max_false_positive.
+    fn new(size: usize, max_false_positive: f64) -> Self {
+        let bit_count = (-1.44 * size as f64 * max_false_positive.log2()).ceil() as usize;
+        let hash_functions = -max_false_positive.log2().ceil() as usize;
+        let bits = vec![false; bit_count];
+
+        Self { bits, hash_functions }
+    }
+
+    /// Inserts the item into the BloomFilter
+    fn insert<T: Hash>(&mut self, item: &T) {
+        for i in 0..self.hash_functions {
+            let ix = self.get_index(item, i);
+            self.bits[ix] = true;
+        }
+    }
```

Comment:
Suggestion: you could try using immutable data structures (https://github.com/bodil/im-rs)

Reply:
Looks interesting, but I wonder how immutable data structured play with a language like Rust. Rust already has some interesting properties which seems to make mutability less of a problem to work with. Basically in rust for some memory A, you can only have one of the following at a time:
* A single mutable reference to A
* Any number of immutable references to A

This means that it is impossible to have data races for instance, and rust will complain at compile time if you don't follow these rules.

But there are probably other instances where this kind of immutable data structures can still be useful.

Reply:
Oh very interesting!
Yeah aside from distributed computing and data race concerns, immutability also promotes a more "pure" style of programming, in the FP sense of the word.
Meaning that your function can only return stuff, but cannot produce side effects. Stuff in, stuff out, and that's it. It makes code much easier to read and test.

Reply:
In VIS, you'd often see stuff like this:
```
... lots of code
this.processCustomMetrics(rows, stuff, foo, bar)
... more code
```
This function doesn't return anything? What does it do? What does it operate on? It must produce side effects, but what?
This is literally made impossible by immutability.